Make YlmSpherepack threadsafe

markscheel opened the pull request:
## Proposed changes

Uses thread_local storage to make YlmSpherepack threadsafe.  Fixes #3929

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
YlmSpherePackHelper: MemoryPool: add clear() function.

markscheel committed:
YlmSpherepack: make threadsafe using thread_local.

nilsdeppe reviewed (commented):
LGTM, I think this should fix the issue :)

nilsvu commented on src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp:
```diff
@@ -437,13 +448,14 @@ class YlmSpherepack {
   size_t spectral_size_;
-  mutable YlmSpherepack_detail::MemoryPool memory_pool_;
-  mutable YlmSpherepack_detail::Storage storage_;
+  static thread_local YlmSpherepack_detail::MemoryPool memory_pool_;
```
What happens when multiple YlmSpherepack objects are around? Won't they all use the same memory pool since it's static?

Can these scenarios be unit-tested?

Also, I think some more code comments and/or dox explaining the design choices here would be helpful.

markscheel commented:
Added code comments addressing @nilsvu 's concern.

Also removed the destructor.  I realized that having the destructor as it was would just increase the number of memory allocations but (probably) would not save much memory.

markscheel commented:
Rebased on develop.

nilsdeppe commented on src/NumericalAlgorithms/SphericalHarmonics/YlmSpherepackHelper.cpp:
```diff
@@ -167,4 +167,13 @@ void MemoryPool::free(const gsl::not_null<double*> to_be_freed) {
 
+void MemoryPool::clear() {
+  for (auto& elem : memory_pool_) {
+    if(elem.currently_in_use) {
```
Is this `clear()` function really not being tested? If so, could you add (or extend) the tests to cover it?

markscheel replied:
Ok, I did that and even more.   My new commit changes all the error tests in Test_YlmSpherepack into CHECK_THROWS_WITH.

nilsdeppe reviewed (commented):
Overall LGTM, just one question

markscheel committed:
Test_YlmSpherepack: error tests -> CHECK_THROWS_WITH.

Also expanded the number of tests for memory_pool.

nilsvu commented:
@markscheel since you mentioned this at the meeting today, I don't have the time to review this at the moment, so please don't wait for my review here. Maybe @wthrowe or @nikwit could provide the second review.

wthrowe commented:
Looks good, and I think your change addresses @nilsvu's request.  Squash.
